hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReader.java

```diff
@@ -267,13 +273,20 @@ public Path getCurrentPath() {
   }
 
   // returns false if we've already exceeded the global quota
-  private boolean checkBufferQuota() {
+  private void blockUntilFreeBufferQuota() {
     // try not to go over total quota
-    if (!this.getSourceManager().checkBufferQuota(this.source.getPeerId())) {
-      Threads.sleep(sleepForRetries);
-      return false;
+    long current = EnvironmentEdgeManager.currentTime();
+    while (
+      !this.getSourceManager().checkBufferQuota(this.source.getPeerId()) && isReaderRunning()
+    ) {
+      if (EnvironmentEdgeManager.currentTime() - current >= logQuotaThrottleInterval) {
+        LOG.warn(
+          "peer={}, source reader check buffer quota failed, current wal is {}, will sleep {}ms for next retry",
+          this.source.getPeerId(), this.getCurrentPath(), sleepForQuotaCheck);
+      }
+      Threads.sleep(sleepForQuotaCheck);
+      current = EnvironmentEdgeManager.currentTime();
```

Comment:
We will reset current everytime here, so the `EnvironmentEdgeManager.currentTime() - current >= logQuotaThrottleInterval` condition will never be met?

Reply:
Yes, thanks, it's an obvious bug.